psi4/src/psi4/scfgrad/jk_grad.cc

```diff
@@ -1529,23 +1529,22 @@ void DFJKGrad::compute_hessian()
                 for (int p = oP; p < oP+nP; p++) {
                     for (int m = oM; m < oM+nM; m++) {
                         for (int n = oN; n < oN+nN; n++) {
-                            Amnp[p][m*nso+n] += (*buffer++);
+                            Amnp[p][m*nso+n] = (*buffer++);
                         }
                     }
                 }
-                // c[A] = (A|mn) D[m][n]
-                C_DGEMV('N', np, nso*(size_t)nso, 1.0, Amnp[0], nso*(size_t)nso, Dtp[0], 1, 0.0, cp, 1);
-                // (A|mj) = (A|mn) C[n][j]
-                C_DGEMM('N','N',np*(size_t)nso,na,nso,1.0,Amnp[0],nso,Cap[0],na,0.0,Amip[0],na);
-                // (A|ij) = (A|mj) C[m][i]
-                #pragma omp parallel for
-                for (int p = 0; p < np; p++) {
-                    C_DGEMM('T','N',na,na,nso,1.0,Amip[p],na,Cap[0],na,0.0,&Aijp[0][p * (size_t) na * na],na);
-                }
-
             }
         }
     }
+    // c[A] = (A|mn) D[m][n]
```

Comment:
I don't quite follow how this is the same, the top looks incorrect as you were repeatedly summing a continuously larger non-zero part of the `Amn` matrix which should give bad results (and fail tests). The new one looks correct however.

Reply:
It's initially transforming a matrix of mostly zeros and *assigning* the result (not accumulating, note the 0.0 arg for beta in the DGEMM and DGEMV calls).  Then during each pass through the integrals, that Amn array becomes more populated, until the final iteration when it's the full TEI batch.  Only then should it be transformed, but the transform was accidentally being performed after each shell computation, overwriting the previously computed transformed ints.  So only the final iteration's transform is actually recorded and the code was correct (but very bad).

Reply:
I just closed the braces in the wrong place, but didn't pick up on it because the answer is still correct and I haven't done any profiling to optimize this code (note it's still in core and has only minimal threading)

Reply:
Makes sense, I see now!